### PR TITLE
feat: add terramate.config.run.env

### PR DESCRIPTION
**Pull request by katcipis**
This introduces the basics for the terramate.config.run.env block parsing. But still doens't expose the feature + there is no hierarchical behavior/merging. This will be addressed on follow up PRs.

### CODE REVIEW

File: hcl/attribute.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import "github.com/hashicorp/hcl/v2/hclsyntax"
+
+// Attribute represents a parsed attribute.
+type Attribute struct {
+	origin string
+	val    *hclsyntax.Attribute
+}
+
+// Attributes represents multiple parsed attributes.
+// The attributes can be sorted by lexicographically by their name.
```
Comment (katcipis) [inline hcl/attribute.go:26]:
```suggestion
// The attributes can be sorted lexicographically by their name.
```

File: hcl/attribute.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hcl
+
+import "github.com/hashicorp/hcl/v2/hclsyntax"
+
+// Attribute represents a parsed attribute.
+type Attribute struct {
+	origin string
```
Comment (katcipis) [inline hcl/attribute.go:21]: Origin info will be good when we eval, so we can provide accurate data about where the original attribute came from.

File: hcl/attribute.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	val    *hclsyntax.Attribute
+}
+
+// Attributes represents multiple parsed attributes.
+// The attributes can be sorted lexicographically by their name.
+type Attributes []Attribute
```
Comment (katcipis) [inline hcl/attribute.go:27]: Since we augment the attribute to have the origin I already added an Attributes type that is sorteable (so we can always have deterministic output etc).

File: cmd/terramate/cli/project.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ type project struct {
 }
 
 func (p project) gitcfg() *hcl.GitConfig {
-	return p.rootcfg.Terramate.RootConfig.Git
+	return p.rootcfg.Terramate.Config.Git
```
Comment (katcipis) [inline cmd/terramate/cli/project.go:42]: since these parsed structure doens't necessarily represents a root config, but any config block on any part of the hierarchy, I renamed it. I think we can also rename the types to reflect that, but will do that in a separate PR because the changeset is quite big.

Comment (i4ki) [reply cmd/terramate/cli/project.go:42]: what about the issue we were talking earlier today? we talked about failing if the user has `terramate.config` in a non-root directory.. but the `run.env` feature kinda requires that.

Comment (katcipis) [reply cmd/terramate/cli/project.go:42]: `terramate.config` is not a root configuration. The root configurations are:

* terramate.required_version
* terramate.config.git

At least as documented today. If everything inside `terramate.config` is supposed to be root only then we need to come up with a different place to put the run/env configurations.

Comment (katcipis) [reply cmd/terramate/cli/project.go:42]: Maybe we should have a project_config block to make it explicit that it has project wide configuations that can only be defined on the top level/root dir ? :thinking:

File: hcl/hcl.go [hunk 5]
```diff
@@ -1154,13 +1202,40 @@ func (p *TerramateParser) parseTerramateSchema() (Config, error) {
...
+		config.Terramate.Config.Run.Env == nil {
+		return nil
+	}
+
+	errs := errors.L()
+	attrs := map[string]Attribute{}
+
+	for _, attr := range config.Terramate.Config.Run.Env.Attributes {
+		name := attr.Value().Name
+		if _, ok := attrs[name]; ok {
```
Comment (i4ki) [inline hcl/hcl.go:1227]: isn't this ensured by the hcl language spec and also validated by hcl lib?

got this in the playground:
```
playground:13,4-5: Attribute redefined; The argument "a" was already set at playground:12,4-5. Each argument may be set only once.
```

File: hcl/hcl.go [hunk 6]
```diff
@@ -1302,3 +1377,21 @@ func listTerramateDirs(dir string) ([]string, error) {
...
 }
+
+func sortAttributes(attrs hclsyntax.Attributes) []*hclsyntax.Attribute {
+	names := make([]string, 0, len(attrs))
+
+	for name := range attrs {
+		names = append(names, name)
+	}
+
+	log.Trace().Str("action", "sortedAttributes()").Msg("Sort attributes.")
```
Comment (i4ki) [inline hcl/hcl.go:1388]:
```suggestion
	log.Trace().Str("action", "sortAttributes()").Msg("Sort attributes.")
```

**Review by i4ki — APPROVED** [review]
lgtm but left some comments to discuss.
Also, I know there's some discussion around the spec going on in other channels.

File: hcl/hcl.go (hunk 5 above)
Comment (katcipis) [reply hcl/hcl.go:1227]: Not ensured if they come from different blocks or different files. Like this is allowed:

```
block {
  a = 666
}

block {
  a = 666
}
```

Here we detect conflicts accross multiple blocks/files, so I think we need this (you can check the tests for these errors too).

File: hcl/hcl_cfg_run_test.go [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/madlambda/spells/assert"
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/hcl"
+)
+
+func TestHCLParserConfigRun(t *testing.T) {
+
+	runEnvCfg := func(hcldoc string) hcl.Config {
+		// Comparing attributes/expressions with hcl/hclsyntax is impossible
```
[the remark this answers is not in the document]
Comment (katcipis) [reply hcl/hcl_cfg_run_test.go:31]: I would love something better for this...let me know if you have any ideas :-(

File: hcl/hcl.go (hunk 5 above)
Comment (i4ki) [reply hcl/hcl.go:1227]: gotcha

File: cmd/terramate/cli/project.go (hunk 4 above)
Comment (katcipis) [reply cmd/terramate/cli/project.go:42]: we discussed the following options with the team:

* Only allow run/env on root and update specs (but if we want hierarchical in the future the problem comes back)
* Handle as project root terramate.required_version OR terramate.config.git instead of overall existence of terramate.config
* Move terramate.config.run.env to some other place not inside terramate.config and keep it as is
* Move `terramate.config` to something like terramate.project to make it more explicit that it is a project wide/root only config
* Make terramate.config a non-root config that is hierarchical + change project root detection

We ended up deciding for "Only allow run/env on root and update specs" for now. So `terramate.config.run.env` will be a root only configuration for now, no hierarchical/merge behavior.

Comment (i4ki) [reply cmd/terramate/cli/project.go:42]: awesome.
:rocket:

**Review by i4ki — APPROVED** [review]
